### PR TITLE
Implement `add_row_number` for Database backends, fix primary key inference for SQLite

### LINKED ISSUE #6921: Implement `add_row_number` in the Database backend
Implement [`add_row_number`](https://github.com/enso-org/design/blob/main/epics/basic-libraries/expressions/design.md?rgh-link-date=2023-05-25T18%3A24%3A07Z#tableadd_row_number) for the Database backend, mirroring the behaviour of in-memory.

The ordering of rows (if unspecified) should be determined by the table's "default ordering". If the "default ordering" does not exist either, the ordering may be non-deterministic or the implementation is free to fail with an error, requesting a specific ordering.

> The `default ordering` is a concept that needs to be clarified and defined better for our tables. It will default to the primary key (if it exists) and can be then modified by `order_by` operations.

### LINKED ISSUE #7037: Primary Key is not read for SQLite temporary tables
The method `get_primary_key` fails to find the primary key for temporary tables in the SQLite backend. This helper is needed to make `default_key_columns` work which is used as the source of defaults for `update_database_table`. Thus, currently, the defaults fail for `update_database_table` in SQLite when the table is temporary.

This is because the SQLite JDBC driver's `getPrimaryKeys` queries `sqlite_schema` to find the primary keys but the temporary tables 'live' in `temp.sqlite_schema` instead, so it is not found. It contains a fallback code to read `pragma table_info(name)` which is the best source for this data - but the code has some bug that makes this code never reached (and another bug which results in it returning only the last key, not all).

As part of this task, once this method is more robust, we could move it to be a public method on connection `find_primary_key`. We can also consider adding a `primary_key` method on `Table` - but we need to think through what will be the implications of calling `primary_key` on modified tables (1).

> 1. For example, what if columns that were part of the source table's primary key were removed? Or modified? Then the original primary key is no longer valid for such modified table. So we may need to think if we really want to add the method or just keep the `Connection` helper. If we want to, it probably should work only for 'simple' tables with 'trivial' context and throw `Illegal_State` for non-trivial contexts.

To implement this, we need to make the logic for getting primary keys dialect-specific (so moving `get_primary_key` into connection and then delegating to the dialect) so that we can add custom logic for SQLite. We could try to reproduce the initial logic from the driver and `UNION` `sqlite_schema` with `temp.sqlite_schema`, but it seems that instead of that, the most robust and simplest way for this will be to instead rely on the `table_info` pragma. We can read its results through: `connection.read_statement 'pragma table_info("table name")'` and check for the `pk` flag.

### PR DESCRIPTION
**Pull request by radeusgd**
### Pull Request Description

Closes #6921 and also closes #7037 

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

File: distribution/lib/Standard/Database/0.0.0-dev/src/Data/Table.enso [hunk 1]
```diff
@@ -559,10 +561,50 @@ type Table
...
+    add_row_number self (name:Text = "Row") (from:Integer = 1) (step:Integer = 1) group_by=[] order_by=[] on_problems=Problem_Behavior.Report_Warning =
+        problem_builder = Problem_Builder.new error_on_missing_columns=True
+        grouping_columns = self.columns_helper.select_columns_helper group_by True problem_builder
+        grouping_columns.each internal_column->
+            column = self.make_column internal_column
+            if column.value_type.is_floating_point then
+                problem_builder.report_other_warning (Floating_Point_Equality.Error column.name)
+        ordering = Table_Helpers.resolve_order_by self.columns order_by problem_builder
+        problem_builder.attach_problems_before on_problems <|
+            order_descriptors = case ordering.is_empty of
```
Comment (GregoryTravis) [inline distribution/lib/Standard/Database/0.0.0-dev/src/Data/Table.enso:574]: Would it be possible to call `self.order_by` and then recursively call `add_row_number`?

File: distribution/lib/Standard/Database/0.0.0-dev/src/Data/Table.enso [hunk 2]
```diff
@@ -559,10 +561,50 @@ type Table
...
+               passed to the particular operation builder. So at this stage
+               there is no way to distinguish the arguments.
+               This is yet another argument for the IR redesign.
+            separator = SQL_Expression.Literal "<!EXPRESSION_SEPARATOR!>"
+            # The SQL row_number() counts from 1, so we adjust the offset.
+            offset = from - step
+            params = [SQL_Expression.Constant offset, SQL_Expression.Constant step] + order_descriptors + [separator] + grouping_expressions
+            new_expr = SQL_Expression.Operation "ROW_NUMBER" params
+
+            type_mapping = self.connection.dialect.get_type_mapping
```
Comment (GregoryTravis) [inline distribution/lib/Standard/Database/0.0.0-dev/src/Data/Table.enso:590]: This could be a helper function, I've seen it in a few places.

**Review by GregoryTravis — APPROVED** [review]

File: distribution/lib/Standard/Database/0.0.0-dev/src/Data/Table.enso (hunk 1 above)
Comment (radeusgd) [reply distribution/lib/Standard/Database/0.0.0-dev/src/Data/Table.enso:574]: That is allowed yes.

Note that
```
t1 = t0.add_row_number order_by=["X"]
t2 = t0.order_by ["X"] . add_row_number
```
has slightly different semantics.

If `t0` is sorted according to some column `Y`, then in `t1` resulting rows will still be in the `Y` order, just the added column numbering will be according to `X`; while in `t2` the rows will be re-ordered to `X`.

File: distribution/lib/Standard/Database/0.0.0-dev/src/Data/Table.enso (hunk 2 above)
Comment (radeusgd) [reply distribution/lib/Standard/Database/0.0.0-dev/src/Data/Table.enso:590]: Could you clarify what exactly?

I don't see what pattern you are referring to.

File: distribution/lib/Standard/Database/0.0.0-dev/src/Data/Table.enso [hunk 3]
```diff
@@ -23,6 +23,7 @@ import Standard.Table.Data.Table.Table as Materialized_Table
 import Standard.Table.Data.Type.Value_Type_Helpers
 import Standard.Table.Internal.Aggregate_Column_Helper
 import Standard.Table.Internal.Table_Helpers
+import Standard.Table.Internal.Add_Row_Number
```
Comment (jdunkerley) [inline distribution/lib/Standard/Database/0.0.0-dev/src/Data/Table.enso]: Can we put in alphabetical order here please.

File: distribution/lib/Standard/Database/0.0.0-dev/src/Data/Table.enso (hunk 1 above)
Comment (jdunkerley) [reply distribution/lib/Standard/Database/0.0.0-dev/src/Data/Table.enso:574]: That is slightly database dependent - t1 can be reordered if no order by clause was specified.
Postgres will for example (as more efficient for it).
If multiple functions in SQL with order by clauses it tends to pick the first one I think.
Obv table order by dominates column.

File: distribution/lib/Standard/Database/0.0.0-dev/src/Internal/Base_Generator.enso [hunk 4]
```diff
@@ -265,6 +266,21 @@ make_is_in_column arguments = case arguments.length of
...
     _ -> Error.throw <| Illegal_State.Error ("The operation IS_IN_COLUMN requires at exactly 3 arguments: the expression, the IN subquery, the subquery checking for nulls.")
 
+## PRIVATE
+make_row_number : Vector Builder -> Builder
+make_row_number arguments = if arguments.length < 4 then Error.throw (Illegal_State.Error "Wrong amount of parameters in ROW_NUMBER IR. This is a bug in the Database library.") else
+    offset = arguments.at 0
+    step = arguments.at 1
+
+    separator_ix = arguments.index_of code->
+        code.build.prepare.first == "<!EXPRESSION_SEPARATOR!>"
```
Comment (jdunkerley) [inline distribution/lib/Standard/Database/0.0.0-dev/src/Internal/Base_Generator.enso]: Lets make this commented SQL just for the sake of safety `--<!EXPRESSION_SEPARATOR!>--`

**Review by jdunkerley — APPROVED** [review]
Couple of tiny comments but looks great :)

File: distribution/lib/Standard/Database/0.0.0-dev/src/Data/Table.enso (hunk 1 above)
Comment (radeusgd) [reply distribution/lib/Standard/Database/0.0.0-dev/src/Data/Table.enso:574]: > That is slightly database dependent - t1 can be reordered if no order by clause was specified.

That's what I meant by:

> If `t0` is sorted according to some column `Y`

i.e. I assumed that `t0` has `ORDER BY Y`. And my point was that `add_row_number` itself will _not override_ this ordering, whereas an `order_by` followed by `add_row_number` will change it.

File: distribution/lib/Standard/Database/0.0.0-dev/src/Internal/Base_Generator.enso (hunk 4 above)
Comment (radeusgd) [reply distribution/lib/Standard/Database/0.0.0-dev/src/Internal/Base_Generator.enso]: Ok. I also moved it into a separate variable so the constant is not duplicated.

**Pull request merged**